parlai/tasks/wizard_of_internet/mutators.py

```diff
@@ -0,0 +1,255 @@
+class AddLabel(AddLabelWizWiki):
+    """
+    Adds the dialogue sentence to the input.
+
+    E.g. run with: parlai display_data -t wizard_of_internet -n 100 -dt valid --mutators
+    flatten,checked_sentence_as_label_woi,add_label_to_input_woi
+    """
+
+    pass
+
+
+@register_mutator("woi_add_label_to_input_lm")
+class AddLabelLM(AddLabelLMWizWiki):
+    """
+    Adds the dialogue sentence to the input (language modeling version).
+
+    Language modeling version where a random piece of the label is sampled in
+    the input. The rest is placed inside special tokens.
+
+    E.g. run with: parlai display_data -t wizard_of_internet -n 100 -dt valid --mutators
+    flatten,add_label_to_input_lm_woi
+
+    To add the checked sentence as the label, use:
+        parlai display_data -t wizard_of_internet -n 100 -dt valid --mutators
+        flatten,add_label_to_input_lm_woi,checked_sentence_as_label_woi
+    """
+
+    pass
+
+
+@register_mutator("woi_filter_no_passage_used")
+class WoiFilterNoPassageUsed(ManyEpisodeMutator):
+    """
+    Allows to filter any examples where no passage was selected to base the wizard reply
+    on.
+
+    This works best in flattened mode. E.g. run with: parlai display_data -t
+    wizard_of_internet -n 100 -dt valid --mutators flatten+filter_no_passage_used
+    """
+
+    def many_episode_mutation(self, episode):
+        out_episodes = []
+        for e in episode:
+            checked_sentences = e.get(CONST.SELECTED_SENTENCES)
+            checked_sentences = ' '.join(checked_sentences)
+            if checked_sentences == CONST.NO_SELECTED_SENTENCES_TOKEN:
+                pass
+            else:
+                out_episodes.append([e])
+        return out_episodes
+
+
+@register_mutator("woi_filter_selected_knowledge_in_retrieved_docs")
+class WoiFilterSelectedKnowledgeInRetrievedDocs(ManyEpisodeMutator):
+    """
+    Allows to filter any examples where '__retrieved-docs__' field doesn't contain the
+    '__selected-sentences__'.
+    """
+
+    def many_episode_mutation(self, episode):
+        out_episodes = []
+        for e in episode:
+            checked_sentences = e.get(
+                CONST.SELECTED_SENTENCES,
+                e.get('labels', [CONST.NO_SELECTED_SENTENCES_TOKEN]),
+            )
+            docs = ' '.join(e.get('__retrieved-docs__'))
+            if ' '.join(checked_sentences) != CONST.NO_SELECTED_SENTENCES_TOKEN:
+                found = True
+                for sent in checked_sentences:
+                    s = sent.lstrip(' ').rstrip(' ')
+                    if s not in docs:
+                        found = False
+                if found:
+                    out_episodes.append([e])
+            else:
+                pass
+        return out_episodes
+
+
+def chunk_docs_in_message(message, chunk_sz):
+    if CONST.RETRIEVED_DOCS not in message:
+        return message
+    new_message = message.copy()
+    docs = message.get(CONST.RETRIEVED_DOCS)
```

Comment:
nit: maybe it's not a bad idea to use `message[CONST.RETRIEVED_DOCS]` to make the code crash here, if this mutator was accidentally used with other teachers. Seeing why it goes wrong later might be a bit harder for someone who doesn't know these mutators to debug.

Reply:
ok